shell/platform/windows/keyboard_key_handler.h

```diff
@@ -0,0 +1,173 @@
+#include "flutter/shell/platform/windows/public/flutter_windows.h"
+#include "rapidjson/document.h"
+
+namespace flutter {
+
+class FlutterWindowsView;
+
+// Handles key events.
+//
+// This class detects whether an incoming event is a redispatched one,
+// dispatches native events to delegates and collect their responses,
+// and redispatches events unhandled by Flutter back to the system.
+// See |KeyboardHook| for more information about dispatching.
+//
+// The exact behavior to handle events are further forwarded into
+// delegates. See |KeyboardKeyHandlerDelegate| and its subclasses.
+class KeyboardKeyHandler : public KeyboardHandlerBase {
+ public:
+  // An interface for concrete definition of how to asynchronously handle key
+  // events.
+  class KeyboardKeyHandlerDelegate {
+   public:
+    // Defines how to how to asynchronously handle key events.
+    //
+    // |KeyboardHook| should invoke |callback| with the response (whether the
+    // event is handled) later for exactly once.
+    virtual void KeyboardHook(int key,
+                              int scancode,
+                              int action,
+                              char32_t character,
+                              bool extended,
+                              bool was_down,
+                              std::function<void(bool)> callback) = 0;
+
+    virtual ~KeyboardKeyHandlerDelegate();
+  };
+
+  using EventRedispatcher =
+      std::function<UINT(UINT cInputs, LPINPUT pInputs, int cbSize)>;
+
+  // Create a KeyboardKeyHandler and specify where to redispatch events.
+  //
+  // The |redispatch_event| is typically |SendInput|.
+  explicit KeyboardKeyHandler(EventRedispatcher redispatch_event);
+
+  ~KeyboardKeyHandler();
+
+  // Add a delegate that handles events received by |KeyboardHook|.
+  void AddDelegate(std::unique_ptr<KeyboardKeyHandlerDelegate> delegate);
+
+  // Handles a key event.
+  //
+  // Returns whether this handler claims to handle the event, which is true if
+  // the event is a native event, or false if the event is a redispatched one.
+  //
+  // Windows requires a synchronous response of whether a key event should be
+  // handled, while the query to Flutter is always asynchronous. This is
+  // resolved by "redispatching": the response to the native event is always
+  // true. If Flutter later decides not to handle the event, an event is then
+  // synthesized, dispatched to system, received again, detected, at which time
+  // |KeyboardHook| returns false, then falls back to other keyboard handlers.
+  //
+  // Received events are further dispatched to all added delegates. If any
+  // delegate returns true (handled), the event is considered handled. When
+  // all delegates responded, any unhandled events are redispatched via
+  // |redispatch_event| and recorded. The next (one) time this exact event is
+  // received, |KeyboardHook| will skip it and immediately return false.
+  //
```

Comment:
You might explain the reasoning behind not allowing a delegate to prevent dispatch to other delegates here. E.g. something about the legacy status of all but one of the delegates.

Reply:
Added, thanks.